### PR TITLE
Address msvc x64 compile warnings

### PR DESCRIPTION
**Pull request by Brett208**
Sorry, this turned into a bit of a monster changeset.

I wanted to start using index over fileIndex and item over file for archives since we are allowing pulling data out of the archive that does not represent an entire file.

I found some odd omissions and mistakes that are now fixed:
 - BitStream.h did not contain a #pragma once
 - There were 2 leftover function prototypes in VolFile and ClmFile from when I refactored the functions into the base class
 - There was an unused VOL_WRITE_SIZE define in VolFile.h

There is still one remaining warning originating from a function in ArchiveUnpacker.

```
// Returns -1 if internalFilename is not present in archive.
int GetInternalItemIndex(const std::string& internalName);
```
I would like to propose changing this to returning a size_t and throwing a runtime_error if the internalName is not found. This would allow unifying the concept of an index in the Archives code as being a size_t. I was using the -1 to prevent raising an exception if the name was not found. However, I think it is not in line with how the rest of the library works now and should be changed.

Thoughts?

-Brett

### CODE REVIEW

File: src/Archives/BitStream.h [hunk 1]
```diff
@@ -1,24 +1,27 @@
...
-		int m_BufferBitSize;		// Size of buffer in bits
-		int m_BufferSize;			// Size of buffer in bytes
+		std::size_t m_BufferBitSize;// Size of buffer in bits
+		std::size_t m_BufferSize;	// Size of buffer in bytes
 		unsigned char *m_Buffer;	// Byte array of data
 
-		int m_ReadBitIndex;			// Current bit being read from the stream
-		int m_WriteBitIndex;		// Current bit being written to the stream
+		std::size_t m_ReadBitIndex;	// Current bit being read from the stream
+		std::size_t m_WriteBitIndex;// Current bit being written to the stream
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/BitStream.h]: I'm not really following what a better match would be. Are you trying to say that a max buffer size of size_t would not allow for bit_size to be stored in the same size variable?

File: src/Archives/BitStream.h [hunk 2]
```diff
@@ -1,24 +1,27 @@
...
 	public:
 		BitStream();				// Construct empty stream
-		BitStream(int bufferSize, void *buffer);// Construct stream around given buffer
+		BitStream(std::size_t bufferSize, void *buffer); // Construct stream around given buffer
 
 		bool ReadNextBit();			// Get bit at Read index and advance index
 		int  ReadNext8Bits();		// Get next 8 bits at Read index and advance index
 		bool EndOfStream() const;	// Returns true if the last bit has been read
-		int  GetBitReadPos() const; // Returns the position (in bits) of the read pointer
+		std::size_t GetBitReadPos() const; // Returns the position (in bits) of the read pointer
```
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Archives/BitStream.h]: Yes. A `std::size_t` can represent any valid byte aligned address for the current architecture. With 8 bits to a byte, there are 8x more possible bit addresses. Hence `std::size_t` is too small to handle the max theoretical buffer size with bit granularity.

File: src/Archives/BitStream.cpp [hunk 3]
```diff
@@ -93,7 +93,7 @@ namespace Archives
 		return m_ReadBitIndex >= m_BufferBitSize;
 	}
 
-	int BitStream::GetBitReadPos() const
+	std::size_t BitStream::GetBitReadPos() const
```
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Archives/BitStream.cpp]: That's not a terrible idea. Thought for 64-bit builds, it's still the same issue, since an `std::size_t` is likely going to be the same size as a `uint64_t` there.

I think in practice, we could just check and throw an exception. If we stuck with `std::size_t`, this means on 32-bit, max array size is 4GB, but max bit stream length is 512MB (4GB / 8). That's still plenty large. On 64-bit, max array size is 2^64 bytes, with max bit stream length 2^64 / 8 = 2^61 bytes.

I would be happy with a constructor check that ensures buffer size is capped at `std::numeric_limits<std::size_t>::max() / 8`, and stuck with using `std::size_t`.

Note: The above [`numeric_limits`](https://en.cppreference.com/w/cpp/types/numeric_limits/max) template is from `#include <limits>`

File: src/Archives/BitStream.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Archives/BitStream.h]: There isn't really a good match. That's the problem.

We might want to use a `typedef` for bit indexes to make it clear they are a little different. For efficiency reasons, the typedef should probably be set to `std::size_t`, with a few sanity checks inserted to guard against overflow.

**Comment by DanRStevens** [comment]
Ok, the idea I had for templated length prefixed vector write compiles. We don't have unit tests to easily check if it works though. It relates to changes here, so I'm including it here, though perhaps this should be its own branch.

I have not yet added corresponding read code.

Side note: We can use this same technique for writing/reading length prefixed strings.

The bounds checking casting part could reasonably be pulled out into its own template, possibly independent of the stream code.

With the above ideas implemented, many MapWriter helper methods could be reduced to trivial one-liners, or possibly eliminated.

**Comment by Brett208** [comment]
I am a fan of the new write template. As mentioned, we could probably improve the comments and exception method. I'm happy you added the new write here, otherwise we would be refactoring code that would be deleted once it was added anyways. Maybe we can create a branch to fill in the rest of templates and clean them up.

Once this branch is cleaned, I go through a small list of tests to include maps, CLM, VOL, and compressed VOLs.

-Brett

File: src/Maps/MapWriter.cpp [hunk 4]
```diff
@@ -82,14 +82,15 @@ namespace MapWriter {
...
+			WriteContainerSize(streamWriter, static_cast<uint32_t>(terrainTypes.size()));
 
 			streamWriter.Write(terrainTypes);
 		}
 
 		void WriteTileGroups(StreamWriter& streamWriter, const std::vector<TileGroup>& tileGroups)
 		{
-			WriteContainerSize(streamWriter, tileGroups.size());
+			// The number of tile groups within the map is stored in a 4 byte integer
+			WriteContainerSize(streamWriter, static_cast<uint32_t>(tileGroups.size()));
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Maps/MapWriter.cpp]: I'd like to move this bounds check into the body of WriteContainerSize in the next commit.

File: src/Archives/BitStream.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/BitStream.cpp]: I would have probably used SIZE_MAX since I wasn't aware of the std::numeric_limits<std::size_t>::max() syntax. Do you prefer one over the other for any reason?

-Brett

File: src/Archives/ClmFile.cpp [hunk 5]
```diff
@@ -270,7 +270,7 @@ namespace Archives
 
 	void ClmFile::PrepareIndex(int headerSize, const std::vector<std::string>& internalNames, std::vector<IndexEntry>& indexEntries)
 	{
-		uint32_t offset = headerSize + internalNames.size() * sizeof(IndexEntry);
+		uint32_t offset = headerSize + static_cast<uint32_t>(internalNames.size()) * sizeof(IndexEntry);
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/ClmFile.cpp]: I question someone's ability to open enough file handles simultaneously to reach this boundary. Perhaps if they have insanely long filenames?

File: src/Archives/HuffLZ.cpp [hunk 6]
```diff
@@ -219,41 +235,65 @@ namespace Archives
...
+		if (offset < 0xC0) {
+			return ((offset - 0x90) >> 2) + 0x0C;
+		}
+		if (offset < 0xF0) {
+			return ((offset - 0xC0) >> 1) + 0x18;
+		}
+
+		return offset - 0xC0;
+	}
+
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/HuffLZ.cpp]: Agreed on the data table. It jumped out to me that the same if/else construct is used twice. From a clean code perspective, we should strive to eliminate this to make future changes easier. Perhaps we could have an int enum represent the 0x20, 0x50, 0x90, 0xC0, 0xF0. We could then use this enum to lookup the needed values from a std::array.

-Brett

File: src/Archives/BitStream.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/BitStream.cpp]: Actually, looking at this a little closer, it doesn't make sense to compare with a max size since it will wrap around and be smaller than max size before comparison. what about:

```
// Check if storing bufferSize in bits wraps past largest value of a std::size_t
if (bufferSize << 3 < bufferSize) {
	throw std::runtime_error("BitStream cannot support a buffer size of " + std::to_string(bufferSize));
}
```

File: src/Archives/VolFile.cpp [hunk 7]
```diff
@@ -282,11 +282,13 @@ namespace Archives
 			volInfo.indexEntries.push_back(indexEntry);
 
 			// Add length of internal filename plus null terminator to string table length.
-			volInfo.stringTableLength += volInfo.internalNames[i].size() + 1;
+			// stringTableLength cannot be greater in size than a 4 byte integer.
+			volInfo.stringTableLength += static_cast<uint32_t>(volInfo.internalNames[i].size()) + 1;
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/VolFile.cpp]: Okay, should be fixed.

File: src/Archives/VolFile.cpp [hunk 8]
```diff
@@ -282,11 +282,13 @@ namespace Archives
...
 		}
 
 		// Calculate size of index table
-		volInfo.indexTableLength = volInfo.fileCount() * sizeof(IndexEntry);
+		// indexTableLength cannot be greater in size than a 4 byte integer
+		volInfo.indexTableLength = static_cast<uint32_t>(volInfo.fileCount()) * sizeof(IndexEntry);
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/VolFile.cpp]: Should be fixed in next set of commits.

File: src/Maps/MapWriter.cpp [hunk 9]
```diff
@@ -115,7 +115,8 @@ namespace MapWriter {
 		// String must be stored in file as string length followed by char[].
 		void WriteString(StreamWriter& streamWriter, const std::string& s)
 		{
-			WriteContainerSize(streamWriter, s.size());
+			// String size must not be greater than a 4 byte integer
+			WriteContainerSize(streamWriter, static_cast<uint32_t>(s.size()));
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Maps/MapWriter.cpp]: Fixed now.

File: src/ResourceManager.cpp [hunk 10]
```diff
@@ -88,11 +88,11 @@ std::vector<std::string> ResourceManager::GetAllFilenamesOfType(const std::strin
 	return filenames;
 }
 
-bool ResourceManager::ExistsInArchives(const std::string& filename, int& volFileIndexOut, int& internalVolIndexOut)
+bool ResourceManager::ExistsInArchives(const std::string& filename, int& volFileIndexOut, size_t& internalVolIndexOut)
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/ResourceManager.cpp]: Agreed. Need to fix function ArchiveUnpacker::GetInternalItemIndex to return a size_t before working on this.

File: src/Archives/ClmFile.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/ClmFile.cpp]: Will be fixed in next commit

**Comment by Brett208** [comment]
I still need to address the int return on ArchiveUnpacker::GetInternalItemIndex and its affect on the ResourceManager function. Since we agree on raising an exception inside GetInternalItemIndex, I'll go ahead with implementing this in the future. It bothers me that the implementation is basically identical to ArchiveUnpacker::ContainsItem, which is maybe part of the reason I hadn't cleaned it up before now. I suppose if we get the function signatures and exception raising correct now then we can come back later and improve the internal implementations on a different branch.

Thanks for the quick and very thorough review of a lot of changes in many different parts of the code base. I think the MSVC warnings pointed out a lot of good changes. It looks like I was generally sloppy in not doing bounds checking before the casts which is hopefully now generally addressed.

I noticed in the BitStream class that the variable m_ReadBuffer was not being set to 0 in either of the constructors. Is this a design feature or oversight? Didn't want to assume an oversight in case it was intentional.

Unfortunately in the commit **Check for BufferBitSize overflow in BitStream**, I accidentally let slip a change in VolFile that should have waited and been committed with **Add bounds checking to static casts in ClmFile and VolFile**. I didn't notice until the commit was public, so I left it alone. Sorry for the mess.

---

I would say a file is something that can be consumed by an existing executable and has a definable extension (such as .png) for Windows. I'm not a sound guy but I'm guessing pushing the PCM data into a file without including the WAVE header information would not fit this definition? 

Using internalFilename and fileIndex everywhere seemed inappropriate since the WAVE extension is stripped from files packed into CLM archives. I'm trying to reduce the confusion to the user since sometimes they will get a full file and sometimes they will not. 

---

I'm quite aware of the differences in the search algorithms between the game and OP2Utility/OP2Archive after some disasters of my own trying to get volume files to work. This is why I want to forbid duplicate names or unusual characters in the filenames as these are where the algorithms will differ in results. It  could save others lots of hair pulling later.

Thanks,
Brett

File: src/Maps/MapWriter.cpp [hunk 11]
```diff
@@ -101,16 +90,19 @@ namespace MapWriter {
...
 
 		// Outpost 2 map files represent container sizes as 4 byte values
-		void WriteContainerSize(StreamWriter& streamWriter, uint32_t size)
+		void WriteContainerSize(StreamWriter& streamWriter, std::size_t size)
 		{
+			if (size > UINT32_MAX) {
+				throw std::runtime_error("Container size is too large for writing into an Outpost 2 maps.");
+			}
+
 			streamWriter.Write(size);
```
Comment (DanRStevens) [inline src/Maps/MapWriter.cpp]: This line has been impacted by the function signature update. This write needs to be updated to include a cast. As it stands, it can write a different amount of data between 32-bit and 64-bit builds.

File: src/Archives/ClmFile.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Archives/ClmFile.cpp]: That's a good point. At minimum this should be documented.

An additional point though, was we'd discussed a some point to not open all the files before processing. In which case, there could be a very large number of files being packed, but they are only opened one at a time during the packing process. That could invalidate the above assumption.

File: src/Archives/BitStream.cpp [hunk 12]
```diff
@@ -1,31 +1,29 @@
...
-		m_BufferSize = bufferSize;
-		m_BufferBitSize = bufferSize << 3;
-		m_Buffer = (unsigned char*)buffer;
-
-		m_ReadBitIndex = 0;
-		m_WriteBitIndex = 0;
-
-		m_WriteBuff = 0;
+		// Check if storing bufferSize in bits wraps past largest value of a std::size_t
+		if (m_BufferBitSize < m_BufferSize) {
```
Comment (DanRStevens) [inline src/Archives/BitStream.cpp]: This check will probably work, but it seems a bit obscure. I think I would prefer:
```cpp
if (bufferSize > std::numeric_limits<typename m_BufferBitSize>::max() / 8)
```

File: src/Archives/VolFile.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Archives/VolFile.cpp]: I think you're right. I waffled back and forth on this one for a while, but now it looks good.

The `.size()` part will be a `std::size_t`, which means on 64-bit, the other values will be promoted to 64-bit before the comparison with `UINT32_MAX`. That means the check happens long before overflow could possibly occur.

On 32-bit, they will all be 32-but unsigned ints, which could potentially overflow before the comparison. In fact, that comparison can never be true for 32-bit. However, memory model limitations on both Windows and Linux should cause an out of memory error before reaching that point. I suppose that's a bit platform specific, but should be good enough.

These overflow considerations with variable sized data types can be really difficult to reason about.

File: src/Archives/HuffLZ.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Archives/HuffLZ.cpp]: I don't know if an `enum` would buy us much, as it doesn't aid lookup, nor lead to any efficiency gains. An enum is mostly for giving a name to specific values.

This is probably not such a high priority item though. We could just ignore it. As written, it works. If we address this at all, we could choose to do so in a future branch.

File: src/ResourceManager.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/ResourceManager.cpp]: I'm not sure I see the relationship.

I think the second parameter should be `std::size_t& volFileIndexOut`, and then remove the cast in:
```cpp
volFileIndexOut = static_cast<int>(i);
```

That would then entail localized data type updates to `ResourceManager::ExtractSpecificFile`.

File: src/Maps/MapWriter.cpp (hunk 11 above)
Comment (Brett208) [reply src/Maps/MapWriter.cpp]: Hmm, okay, I'll update soon.

File: src/Maps/MapWriter.cpp [hunk 13]
```diff
@@ -92,7 +92,7 @@ namespace MapWriter {
 			// The number of tile groups within the map is stored in a 4 byte integer
 			WriteContainerSize(streamWriter, static_cast<uint32_t>(tileGroups.size()));
 
-			uint32_t unknown = tileGroups.size() - 1; // Write unknown field with best guess as to what value it should hold
+			uint32_t unknown = static_cast<uint32_t>(tileGroups.size()) - 1; // Write unknown field with best guess as to what value it should hold
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Maps/MapWriter.cpp]: I'll set it to 0 if there are 0 tileGroups. I must have missed this comment earlier. This unknown value is getting a lot of attention.

File: src/Archives/VolFile.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Archives/VolFile.cpp]: Good point. For consistency, I will cast one of the values to uint64_t to allow the check to occur on 32-bit builds. Even if memory problems would keep you from getting there.

**Comment by Brett208** [comment]
Made the last commits to this branch. This branch has covered a lot of code and GitHub comments. Are there any issues with the newly committed code, or did I miss any fixes from earlier comments that still need to be made?

Oh, any issue with adding m_ReadBuff(0) to constructor chains for 2 constructors in BitStream?

The following plans to remain outstanding until later:
 * Update MapReader to include a Read template for reading a vector of a specific max size
 * Improve comments and possibly variable names on Write template for writing a vector of a specific max size
 * Include a Read/Write template for a std::string of a specific max size
 * Remove repeated if/else statements in HuffLZ GetNumExtraBits & GetRepeatOffset

Thanks,
-Brett

File: src/Archives/BitStream.cpp (hunk 12 above)
Comment (Brett208) [reply src/Archives/BitStream.cpp]: Saying my code appears obscure seems like an odd role reversal. I liked it because it didn't require any division and was shorter :).

Including the limits header with the block of code shown above produces the following error on MSVC:

Error (active) E0260 explicit type is missing ('int' assumed) OP2Utility c:\ . . . \OP2Archive\OP2Utility\src\Archives\BitStream.cpp 26

and

Error (active) E0439 expected a '>'	OP2Utility c:\ . . . \OP2Archive\OP2Utility\src\Archives\BitStream.cpp 26

**Comment by DanRStevens** [comment]
Ok, I might not be able to give this a proper review for a couple of days. Indeed, there is a lot to check over, and too many comments for me to keep track of it all. Thank you for a summary of the outstanding items.

One thing that comes to mind that wasn't mentioned is the File -> Item change. I still feel like "Item" doesn't add any meaningful description to the name. I'm thinking perhaps we should consider shortening method names further. I also think my earlier use of "Internal" was excessive, and should probably be cut. For example, maybe we should use these shorter names for ArchiveUnpacker:
```cpp
bool Contains(const std::string& name);
std::size_t GetIndex(const std::string& name);

virtual std::string GetName(std::size_t index) = 0;
virtual uint32_t GetSize(std::size_t index) = 0;
```

----

There shouldn't be any problems with initializing `m_ReadBuff` in the constructor. As the class is written, initializing it won't do anything, though may be cleaner. The code paths, as controlled by the other fields, which are initialized, ensure `m_ReadBuff` is set before its value can be used to affect outputs.

I had to remind myself how the class worked. The `m_ReadBuff` field contains the bits remaining to be read from the current byte. It's basically a 1 byte buffer used to support bit reads. The stream offset controls reading a new byte into the buffer.

**Comment by Brett208** [comment]
For consistency I pushed a commit that initialized BitStream::m_ReadBuff to 0 in the constructor initialization lists. 

I pushed the suggested archives function and argument name shortening. I'm happy with the changes or with the more verbose use of internal. I also set a couple of names as xxxCount as opposed to NumberOfxxx, as I prefer the count postfix.

Once this branch is merged, we will need to address the public field name changes in OP2Archive and OP2MapImager.

-Brett

File: src/Archives/VolFile.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Archives/VolFile.cpp]: On 64-bit builds, this check could theoretically overflow for really large `volInfo` vectors. This isn't a hugely practical concern though, since any real-world system would run out of memory before reaching that point.

File: src/Archives/ArchiveUnpacker.cpp [hunk 14]
```diff
@@ -8,55 +8,55 @@
 namespace Archives
 {
 	ArchiveUnpacker::ArchiveUnpacker(const std::string& filename) :
-		m_ArchiveFilename(filename), m_NumberOfPackedItems(0), m_ArchiveFileSize(0) { }
+		m_ArchiveFilename(filename), m_PackedItemCount(0), m_ArchiveFileSize(0) { }
```
Comment (DanRStevens) [inline src/Archives/ArchiveUnpacker.cpp]: Maybe rename `m_PackedCount`.

File: src/Archives/ArchiveUnpacker.h [hunk 15]
```diff
@@ -19,23 +19,23 @@ namespace Archives
 
 		std::string GetVolumeFilename() { return m_ArchiveFilename; };
 		uint64_t GetVolumeFileSize() { return m_ArchiveFileSize; };
-		std::size_t GetNumberOfPackedFiles() { return m_NumberOfPackedItems; };
-		bool ContainsItem(const std::string& internalName);
-		std::size_t GetInternalItemIndex(const std::string& internalName);
-		void ExtractFile(const std::string& internalFilename, const std::string& pathOut);
+		std::size_t GetPackedItemCount() { return m_PackedItemCount; };
```
Comment (DanRStevens) [inline src/Archives/ArchiveUnpacker.h]: Maybe rename to `GetPackedCount()`.

File: src/Archives/ArchiveUnpacker.h [hunk 16]
```diff
@@ -19,23 +19,23 @@ namespace Archives
...
+		virtual uint32_t GetSize(std::size_t index) = 0;
 		virtual void ExtractFile(std::size_t index, const std::string& pathOut) = 0;
 		virtual void ExtractAllFiles(const std::string& destDirectory);
 		virtual std::unique_ptr<SeekableStreamReader> OpenStream(std::size_t index) = 0;
-		virtual std::unique_ptr<SeekableStreamReader> OpenStream(const std::string& internalName);
+		virtual std::unique_ptr<SeekableStreamReader> OpenStream(const std::string& name);
 
 	protected:
-		void CheckPackedIndexBounds(std::size_t index);
+		void CheckIndexBounds(std::size_t index);
```
Comment (DanRStevens) [inline src/Archives/ArchiveUnpacker.h]: Yes, actually, I think I like removing "Packed" as well. I suppose "Packed" is roughly equivalent to "Internal", which we were removing.

The previous API allowed for different filenames inside the archive from the source file being packed, which was one reason to distinguish between internal names and external names, or packed files versus unpacked files. Now that the API has been simplified, this distinction no longer seems necessary. Or rather, it is handled by a couple of methods with `name` and `pathOut` to mark the distinction.

**Comment by DanRStevens** [comment]
> I would say a file is something that can be consumed by an existing executable and has a definable extension (such as .png) for Windows.

Not a huge point, but ...

I think I would disagree here. Slightly off topic, but for one, file extensions are a Windows thing. Linux has no special treatment for file extensions, in that there is no OS supported association of file extensions with programs used to open those files. The extension is simply a part of the filename, like any other component, with no special relevance. I'd also say that not having an executable to consume a file makes it no less of a file. That would also open the question of what it means to not have an executable. Anywhere? On Windows? Currently installed? What about one being written, or with plans to be written in the next couple of months? It seems too arbitrary.

I would go with the simple definition that a file is simply a named ordered sequence of bytes.

An additional point, the GoldWave audio file editor supports header-less PCM data files. Sometimes it refers to them as ".raw" files, though the extension is really quite arbitrary and unimportant. It doesn't require any specific extension to process files as such.

That's also exactly what CLM files store, raw PCM data. It just so happens the ".wav" format is more widely used, and more convenient to work with. We just happen to provide data format conversion support during packing/unpacking. The same could be done for the tilesets with ".png" or ".tiff" files, or line ending conversions for the ".txt" files, just there is less incentive to do so there.

In short, I see the world differently on this on, and still consider both to be archives of "files".

File: src/Maps/MapWriter.cpp [hunk 17]
```diff
@@ -74,8 +74,13 @@ namespace MapWriter {
...
 			WriteContainerSize(streamWriter, tileGroups.size());
 
+			uint32_t unknown = 0;
+
 			// tileGroups.size is checked to ensure it is below UINT32_MAX by previous call to WriteContainerSize.
-			uint32_t unknown = static_cast<uint32_t>(tileGroups.size()) - 1; // Write unknown field with best guess as to what value it should hold
+			if (!tileGroups.empty()) {
+				unknown = static_cast<uint32_t>(tileGroups.size()) - 1; // Write unknown field with best guess as to what value it should hold
+			}
+
```
Comment (DanRStevens) [inline src/Maps/MapWriter.cpp]: This looks good. Could possible make it more compact using the ternary operator:
```cpp
// Write unknown field with best guess as to what value it should hold
uint32_t unknown = !tileGroups.empty() ? static_cast<uint32_t>(tileGroups.size()) - 1 : 0;
streamWriter.Write(unknown);
```

I'd also suggest fewer blank lines around the `unknown` field code, so it groups together visually.

File: src/ResourceManager.cpp [hunk 18]
```diff
@@ -155,9 +160,8 @@ std::string ResourceManager::FindContainingArchiveFile(const std::string& filena
 {
 	for (const auto& archiveFile : ArchiveFiles)
 	{
-		int internalFileIndex = archiveFile->GetInternalItemIndex(filename);
-
-		if (internalFileIndex != -1) {
+		if (archiveFile->ContainsItem(filename)) {
+			std::size_t index = archiveFile->GetInternalItemIndex(filename);
```
Comment (DanRStevens) [inline src/ResourceManager.cpp]: Small side note, but this code will do the name lookup twice. Once to see if the file exists, and then another to return the actual index. For code like this, a sentinel return value is more efficient as the lookup can be done once, while the returned value is used twice (does it exist? where does it exist?). For that reason, I often code both a sentinel return value version and an exception raising version of certain methods.

Though this is perfectly correct, and I think we should move on.

File: src/Archives/BitStream.cpp (hunk 12 above)
Comment (DanRStevens) [reply src/Archives/BitStream.cpp]: Hmm, seems I gave you bad code. Guess I should have tested the idea first. The correction is:
```cpp
if (bufferSize > std::numeric_limits<decltype(m_BufferBitSize)>::max() / 8)
```

----

Indeed yes, your code is shorter, and uses very simple language features. My concern is that shorter isn't always easier to understand. In particular, it is an indirect test. It appears to be testing something which shouldn't possibly be true. It works because of overflow, which is not a super obvious thing to rely on. Worth mentioning, overflow is only well defined for `unsigned` values, so you need to know the type to ensure it is correct.

Using `numeric_limits`, the test doesn't rely on overflow to work. It's a more direct test as to the max buffer size that can be supported.

File: src/Archives/ClmFile.cpp [hunk 19]
```diff
@@ -272,16 +272,17 @@ namespace Archives
 	{
 		uint64_t offset = headerSize + names.size() * sizeof(IndexEntry);
 		for (std::size_t i = 0; i < names.size(); ++i)
-		{
-			// Disable MSVC warning 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead.
+		{      
+      // Disable MSVC warning 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead.
```
Comment (DanRStevens) [inline src/Archives/ClmFile.cpp]: This merge commit seems to have introduced some whitespace errors. There are spaces here instead of the usual tabs used throughout the rest of the file.

Other blocks of code follow with the same problem.

**Comment by DanRStevens** [comment]
Ok, I've taken a quick scan through the comments. Hopefully I didn't miss anything.

I've implemented a few of my suggested changes. I removed the last few mentions of "Item", preferring shorter names. I was tempted to remove all occurrences of "Packed" too, but found `GetPackedCount()` seemed too short and non-descriptive without it. Would be curious if you have a better suggestion for that name.

I created issues for the outstanding items.

I think this might be good to merge. Any missed items can be addressed in a future issue/branch.

File: src/ResourceManager.cpp (hunk 18 above)
Comment (Brett208) [reply src/ResourceManager.cpp]: Agree with this comment in general

**Comment by Brett208** [comment]
Thank you for fixing the whitespace differences I introduced. No issues with the introduction of the ternary operator or the name changes, etc. 

I would be fine shortening the name to GetCount. In context it would read as ArchiveFile.GetCount(). I think it is pretty clear that getting the count of an archive would be a count of the number of discrete items stored in the archive.

-Brett

File: src/ResourceManager.cpp (hunk 18 above)
Comment (DanRStevens) [reply src/ResourceManager.cpp]: So I later found out `index` was unused and removed that line. That makes the side note largely irrelevant in this case.

**Comment by DanRStevens** [comment]
Ok, I made the suggested name change. I think this is good to go. Would you like to do the honours?
